README.md
# JCirclize

[![Codacy Badge](https://api.codacy.com/project/badge/Grade/807123315cce4d12a7132181203cf46e)](https://www.codacy.com/app/francescovallone/jCirclize?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=francescovallone/jCirclize&amp;utm_campaign=Badge_Grade)

## Description

Percentage Circles Plugin for [jQuery](https://jquery.com/)

## Installation

Download the [repository][repo].

## Usage

### Load The Plugin

Use `<script src="dist/jquery.jCirclize.js"></script>` to load the jQuery version of the plugin.
*REMEMBER:* You have to load jQuery to use this version.

or use `<script src="dist/vanilla.jCirclize.js"></script>` to load the vanilla version of the plugin.

And use `<link href="css/jCirclize.css" rel="stylesheet" type="text/css">` to load the style inside your HTML page.

*WARNING* You have to load one version of the plugin and the css file to use this plugin.

## Basic Usage

You can just call the function without any options.

~~~javascript

// JQuery Version

$(document).ready(function() {
    $('#object').circlize();
});

// JavaScript Version

var element = document.getElementById("yourElement");
circlize(element, {});


~~~

Calling the function you set the element to the default parameters, it creates something like this:

![Preview](https://i.imgur.com/bqkjhDO.png)

## Default `options`

You can customize the percentage circle using some options of jCirclize. These are the defaults.

~~~javascript

// JQuery Version

$(document).ready(function() {
    $('#object').circlize({
		radius: 100, // The radius of the circle
		percentage: 50, // The percentage covered by the foreground
		usePercentage: true, // If you turn it to false it will use min and max
		useAnimations: true, // If you turn it to false it will not use the animations
		useGradient: true, // If you turn it to false it will use the solid color of the foreground
		background: "rgba(20,20,20,0.5)",
		foreground: "#1a1a1a",
		stroke: 20, // The width of the stroke
		duration: 1000, // The duration of the animation in milliseconds
		min: 100, // The minimum value
		max: 100, // The maximum value
		gradientColors: ["#f0f0f0", "red", "#f0f0f0"] // [Max 3 colors] Here you can set the colors of the gradients
	});
});

// JavaScript Version

var element = document.getElementById("yourElement");
circlize(element, {
	radius: 100, // The radius of the circle
	percentage: 50, // The percentage covered by the foreground
	usePercentage: true, // If you turn it to false it will use min and max
	useAnimations: true,// If you turn it to false it will not use the animations
	useGradient: true, // If you turn it to false it will use the solid color of the foreground
	background: "rgba(20,20,20,0.5)",
	foreground: "#1a1a1a",
	stroke: 20, // The width of the stroke
	duration: 1000, // The duration of the animation in milliseconds
	min: 100, // The minimum value
	max: 100, // The maximum value
	gradientColors: ["#f0f0f0", "red", "#f0f0f0"] // [Max 3 colors] Here you can set the colors of the gradients
});

~~~

This is my first plugin for JQuery, I hope you like it.

[repo]: https://github.com/francescovallone/jCirclize/archive/master.zip


CHANGELOG.md
# Changelog

Releases are be numbered in the semantic versioning format:

`<major>.<minor>.<patch>`

## v1.0.2

Added vanilla version of the library.

## v1.0.1

Added animations.
Added minimum and maximum values.

## v1.0.0

First release of jCirclize.


src/jCirclize.js
(function( $ ){
	$.fn.circlize = function(options) {
		var defaults = {
			radius: 100,
			percentage: 50,
			usePercentage: true,
			useAnimations: true,
			useGradient: true,
			background: "rgba(20,20,20,0.5)",
			foreground: "#1a1a1a",
			stroke: 20,
			duration: 1000,
			min: 100,
			max: 200,
			gradientColors: ["#f0f0f0", "red", "#f0f0f0"]
		};
		var opts = $.extend(true ,{}, defaults, options );
		return this.each(function() {
			var perc, box, x, y, html, context, cnv, ctn, fore, gradient;
			box = (Math.PI*(opts.radius));
			x = box/2;
			y = box/2;
			perc = opts.usePercentage ? opts.percentage : opts.min;
			html = "<canvas class=\"circle\" width="+box+" height="+box+"></canvas>" + 
						"<canvas class=\"circle\" width="+box+" height="+box+"></canvas>" +
						"<div class=\"percentage\">"+ perc +"</div>";
			$(this).append(html);
			$(this).addClass("canvasized");
			cnv = $(this).children(".circle");
			context = $(cnv)[0].getContext("2d");
			context.translate(0, box);
			context.rotate(-Math.PI / 2);
			if(opts.useGradient){
				gradient = context.createLinearGradient(0,0,opts.radius*Math.PI,0);
				gradient.addColorStop(0, opts.gradientColors[0]);
				gradient.addColorStop(0.5, opts.gradientColors[1]);
				gradient.addColorStop(1, opts.gradientColors[2]);
				fore = gradient;
			}else{
				fore = opts.foreground;
			}
			if(opts.useAnimations){
				jQuery({ counter: 0 }).animate({ counter: $(".percentage").text()}, {
					duration: opts.duration,
					easing: "swing",
					step() {
						if(opts.usePercentage){
							$(".percentage").text(Math.ceil(this.counter) + "%");
							context.beginPath();
							context.arc(x, y, opts.radius, (2-(Math.ceil(this.counter)/100)*2)*Math.PI, 2*Math.PI);
							context.fillStyle = "transparent";
							context.fill();
							context.strokeStyle = fore;
							context.lineWidth = opts.stroke;
							context.stroke();
						}else{
							$(".percentage").text(Math.ceil(this.counter * 10)/10 + "/" + Math.ceil(opts.max * 10)/10);
							context.beginPath();
							context.arc(x, y, opts.radius, (2-(Math.ceil(this.counter)/opts.max)*2)*Math.PI, 2*Math.PI);
							context.fillStyle = "transparent";
							context.fill();
							context.strokeStyle = fore;
							context.lineWidth = opts.stroke;
							context.stroke();
						}
					}
				});
			}else{
				if(opts.usePercentage){
					$(".percentage").text(opts.percentage + "%");
					context.beginPath();
					context.arc(x, y, opts.radius, (2-(Math.ceil(opts.percentage)/100)*2)*Math.PI, 2*Math.PI);
					context.fillStyle = "transparent";
					context.fill();
					context.strokeStyle = fore;
					context.lineWidth = opts.stroke;
					context.stroke();
				}else{
					$(".percentage").text(Math.ceil(opts.min * 10)/10 + "/" + Math.ceil(opts.max * 10)/10);
					context.beginPath();
					context.arc(x, y, opts.radius, (2-(Math.ceil(opts.min)/opts.max)*2)*Math.PI, 2*Math.PI);
					context.fillStyle = "transparent";
					context.fill();
					context.strokeStyle = fore;
					context.lineWidth = opts.stroke;
					context.stroke();
				}
			}
			ctn = $(cnv)[1].getContext("2d");
			ctn.beginPath();
			ctn.arc(x, y, opts.radius, 0*Math.PI, 2*Math.PI);
			ctn.fillStyle = "transparent";
			ctn.fill();
			ctn.strokeStyle = opts.background;
			ctn.lineWidth = opts.stroke;
			ctn.stroke();
		});
	};
})( jQuery );


src/jCirclize-v.js
function circlize(e, options){
    var defaults = {
        radius: 100,
        percentage: 50,
        usePercentage: true,
        useAnimations: true,
        useGradient: true,
        background: "rgba(20,20,20,0.5)",
        foreground: "#1a1a1a",
        stroke: 20,
        duration: 1000,
        min: 100,
        max: 100,
        gradientColors: ["#f0f0f0", "red", "#f0f0f0"]
    };
    var o = Object.assign({}, defaults, options);
    var perc, box, x, y, html, context, cnv, ctn, speed, counter=1, fore, gradient;
    box = (Math.PI*(o.radius));
    x = box/2;
    y = box/2;
    speed = 1000/60;
    perc = o.usePercentage ? o.percentage : o.min;
    html = "<canvas class=\"circle\" width="+box+" height="+box+"></canvas>" + 
            "<canvas class=\"circle\" width="+box+" height="+box+"></canvas>" +
            "<div class=\"percentage\">"+ perc +"</div>";
    e.innerHTML = html;
    e.classList.add("canvasized");
    cnv = e.childNodes;
    context = cnv[0].getContext("2d");
    context.translate(0, box);
    context.rotate(-Math.PI / 2);
    gradient = context.createLinearGradient(0,0,o.radius*Math.PI,0);
    gradient.addColorStop(0, o.gradientColors[0]);
    gradient.addColorStop(0.5, o.gradientColors[1]);
    gradient.addColorStop(1, o.gradientColors[2]);
    fore = o.useGradient ? gradient : o.foreground;
    if(o.useAnimations){
        var timerId = setInterval(function() {
            counter++;
            if(o.usePercentage){
                cnv[2].innerHTML = counter + "%";
                context.beginPath();
                context.arc(x, y, o.radius, (2-(counter/100)*2)*Math.PI, 2*Math.PI);
                context.fillStyle = "transparent";
                context.fill();
                context.strokeStyle = fore;
                context.lineWidth = o.stroke;
                context.stroke();
            }else{
                cnv[2].innerHTML = (counter * 10)/10 + "/" + (o.max * 10)/10;
                context.beginPath();
                context.arc(x, y, o.radius, (2-(counter/o.max)*2)*Math.PI, 2*Math.PI);
                context.fillStyle = "transparent";
                context.fill();
                context.strokeStyle = fore;
                context.lineWidth = o.stroke;
                context.stroke();
            }
            if (counter === perc){
                clearInterval(timerId);
            }
        },speed);
    }else{
        if(o.usePercentage){
            cnv[2].innerHTML = o.percentage + "%";
            context.beginPath();
            context.arc(x, y, o.radius, (2-(o.percentage/100)*2)*Math.PI, 2*Math.PI);
            context.fillStyle = "transparent";
            context.fill();
            context.strokeStyle = fore;
            context.lineWidth = o.stroke;
            context.stroke();
        }else{
            cnv[2].innerHTML = (o.min * 10)/10 + "/" + (o.max * 10)/10;
            context.beginPath();
            context.arc(x, y, o.radius, (2-(o.min/o.max)*2)*Math.PI, 2*Math.PI);
            context.fillStyle = "transparent";
            context.fill();
            context.strokeStyle = fore;
            context.lineWidth = o.stroke;
            context.stroke();
        }
    }
    ctn = cnv[1].getContext("2d");
	ctn.beginPath();
	ctn.arc(x, y, o.radius, 0*Math.PI, 2*Math.PI);
	ctn.fillStyle = "transparent";
	ctn.fill();
	ctn.strokeStyle = o.background;
	ctn.lineWidth = o.stroke;
	ctn.stroke();
}
